// react-billboardjs/es/index.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

// external dependencies
import shallowEqual from 'fbjs/lib/shallowEqual';
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// billboard
import bb from './bb';

// shapes
import { AREA_SHAPE, AXIS_SHAPE, BAR_SHAPE, COLOR_SHAPE, DATA_SHAPE, DONUT_SHAPE, GAUGE_SHAPE, GRID_SHAPE, INTERACTION_SHAPE, LEGEND_SHAPE, LINE_SHAPE, PADDING_SHAPE, PIE_SHAPE, POINT_SHAPE, REGION_SHAPE, RESIZE_SHAPE, SIZE_SHAPE, SPLINE_SHAPE, SUBCHART_SHAPE, SVG_SHAPE, TITLE_SHAPE, TOOLTIP_SHAPE, TRANSITION_SHAPE, ZOOM_SHAPE } from './shapes';

export var createComponentDidMount = function createComponentDidMount(instance) {
  /**
   * @function componentDidMount
   *
   * @description
   * on mount, update the chart based on props
   */
  return function () {
    requestAnimationFrame(function () {
      instance.updateChart(instance.props);
    });
  };
};

export var createShouldComponentUpdate = function createShouldComponentUpdate(instance) {
  /**
   * @function shouldComponentUpdate
   *
   * @description
   * if the component is pure, base the update on whether props and context have changed
   *
   * @param {Object} nextProps the next props
   * @param {Object} nextState the next state
   * @param {Object} nextContext the next context
   * @returns {boolean} should the component update
   */
  return function (nextProps, nextState, nextContext) {
    var isPure = nextProps.isPure;


    return isPure ? !shallowEqual(instance.props, nextProps) || !shallowEqual(instance.context, nextContext) : true;
  };
};

export var createComponentWillUpdate = function createComponentWillUpdate(instance) {
  /**
   * @function componentWillUpdate
   *
   * @description
   * when the component will update, update the chart with the new props
   *
   * @param {Object} nextProps the next props
   */
  return function (nextProps) {
    instance.updateChart(nextProps);
  };
};

export var createComponentWillUnmount = function createComponentWillUnmount(instance) {
  /**
   * @function componentWillUnmount
   *
   * @description
   * prior to unmount, destroy the chart
   */
  return function () {
    instance.destroyChart();
  };
};

export var createAssignElementToRef = function createAssignElementToRef(instance, refName) {
  /**
   * @function assignElementToRef
   *
   * @description
   * set the element DOM node to the refName passed
   *
   * @param {HTMLElement} element the element to assign to the ref
   */
  return function (element) {
    instance[refName] = element;
  };
};

export var createDestroyChart = function createDestroyChart(instance) {
  /**
   * @function destroyChart
   *
   * @description
   * destroy the chart and set the ref to null
   */
  return function () {
    try {
      instance.chart.destroy();

      instance.chart = null;
    } catch (error) {
      console.error('Internal billboard.js error', error); // eslint-disable-line no-console
    }
  };
};

var createGenerateChart = function createGenerateChart(instance) {
  /**
   * @function generateChart
   *
   * @description
   * generate the chart based on the props passed
   *
   * @returns {Object} the chart instance that was generated
   */
  return function () {
    var _instance$props = instance.props,
        classNameIgnored = _instance$props.className,
        isPureIgnored = _instance$props.isPure,
        styleIgnored = _instance$props.style,
        unloadBeforeLoadIgnored = _instance$props.unloadBeforeLoad,
        config = _objectWithoutProperties(_instance$props, ['className', 'isPure', 'style', 'unloadBeforeLoad']);

    return bb().generate(_extends({
      bindto: instance.chartElement
    }, config));
  };
};

export { createGenerateChart };
export var createLoadData = function createLoadData(instance) {
  /**
   * @function loadData
   *
   * @description
   * load new data to the existing chart
   *
   * @param {Object} data the data to load
   */
  return function (data) {
    instance.chart.load(data);
  };
};

export var createRedraw = function createRedraw(instance) {
  /**
   * @function redraw
   *
   * @description
   * trigger a redraw of the chart
   */
  return function () {
    instance.chart.flush();
  };
};

export var createUnloadData = function createUnloadData(instance) {
  /**
   * @function unloadData
   *
   * @description
   * unload data from the existing chart
   *
   * @param {Object} data the data to unload
   */
  return function (data) {
    instance.chart.unload(data);
  };
};

export var createUpdateChart = function createUpdateChart(instance) {
  /**
   * @function updateChart
   *
   * @description
   * update the chart with the new data
   *
   * @param {Object} props the props to update the chart with
   */
  return function (props) {
    var data = props.data,
        unloadBeforeLoad = props.unloadBeforeLoad;


    if (!instance.chart) {
      instance.chart = instance.generateChart(props);
    }

    if (unloadBeforeLoad) {
      instance.unloadData();
    }

    instance.loadData(data);
  };
};

var BillboardChart = function (_Component) {
  _inherits(BillboardChart, _Component);

  function BillboardChart() {
    var _temp, _this, _ret;

    _classCallCheck(this, BillboardChart);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _Component.call.apply(_Component, [this].concat(args))), _this), _this.componentDidMount = createComponentDidMount(_this), _this.shouldComponentUpdate = createShouldComponentUpdate(_this), _this.componentWillUpdate = createComponentWillUpdate(_this), _this.componentWillUnmount = createComponentWillUnmount(_this), _this.chart = null, _this.chartElement = null, _this.destroyChart = createDestroyChart(_this), _this.generateChart = createGenerateChart(_this), _this.loadData = createLoadData(_this), _this.redraw = createRedraw(_this), _this.setChartRef = createAssignElementToRef(_this, 'chartElement'), _this.unloadData = createUnloadData(_this), _this.updateChart = createUpdateChart(_this), _temp), _possibleConstructorReturn(_this, _ret);
  }

  // lifecycle methods


  // instance values


  // instance methods


  BillboardChart.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        style = _props.style;


    return (
      /* eslint-disable prettier */
      React.createElement('div', {
        className: className,
        ref: this.setChartRef,
        style: style
      })
      /* eslint-enable prettier */

    );
  };

  return BillboardChart;
}(Component);

BillboardChart.displayName = 'BillboardChart';
BillboardChart.propTypes = {
  area: AREA_SHAPE,
  axis: AXIS_SHAPE,
  bar: BAR_SHAPE,
  className: PropTypes.string,
  clipPath: PropTypes.bool,
  color: COLOR_SHAPE,
  data: DATA_SHAPE.isRequired,
  donut: DONUT_SHAPE,
  gauge: GAUGE_SHAPE,
  grid: GRID_SHAPE,
  interaction: INTERACTION_SHAPE,
  isPure: PropTypes.bool,
  legend: LEGEND_SHAPE,
  line: LINE_SHAPE,
  onafterinit: PropTypes.func,
  onbeforeinit: PropTypes.func,
  oninit: PropTypes.func,
  onmouseout: PropTypes.func,
  onmouseover: PropTypes.func,
  onrendered: PropTypes.func,
  onresize: PropTypes.func,
  onresized: PropTypes.func,
  padding: PADDING_SHAPE,
  pie: PIE_SHAPE,
  point: POINT_SHAPE,
  regions: PropTypes.arrayOf(REGION_SHAPE),
  resize: RESIZE_SHAPE,
  size: SIZE_SHAPE,
  spline: SPLINE_SHAPE,
  style: PropTypes.object,
  subchart: SUBCHART_SHAPE,
  svg: SVG_SHAPE,
  title: TITLE_SHAPE,
  tooltip: TOOLTIP_SHAPE,
  transition: TRANSITION_SHAPE,
  unloadBeforeLoad: PropTypes.bool,
  zoom: ZOOM_SHAPE
};
BillboardChart.defaultProps = {
  isPure: false,
  unloadBeforeLoad: false
};


export default BillboardChart;
